services/app-api/utils/reports/reports.ts

```diff
@@ -0,0 +1,76 @@
+import { getPossibleFieldsFromFormTemplate } from "../formTemplates/formTemplates";
+import s3Lib, { getFieldDataKey } from "../s3/s3-lib";
+import { AnyObject, State } from "../types";
+
+/**
+ *
+ * @param reportBucket bucket name
+ * @param state state
+ * @param copySourceId fieldDataId of source report
+ * @param formTemplate form template json object
+ * @param validatedFieldData validated field data from request
+ */
+export async function copyFieldDataFromSource(
+  reportBucket: string,
+  state: string | undefined,
+  copySourceId: any,
+  formTemplate: any,
+  validatedFieldData: AnyObject
+) {
+  const sourceFieldData = (await s3Lib.get({
+    Bucket: reportBucket,
+    Key: getFieldDataKey(state as State, copySourceId),
+  })) as AnyObject;
+
+  if (sourceFieldData) {
+    const possibleFields = getPossibleFieldsFromFormTemplate(formTemplate);
+    Object.keys(sourceFieldData).forEach((key: string) => {
+      // Only iterate through entities, not choice lists
+      if (Array.isArray(sourceFieldData[key])) {
+        pruneEntityData(
+          sourceFieldData,
+          key,
+          sourceFieldData[key],
```

Comment:
Open question: Is there a reason sourceFieldData and sourceFieldData[key] are passed in here, rather than sourceFieldData, and a key, and letting the pruneEntityData know how to navigate that?

Do we see it always operating on an object and a known index in that object, or are we preparing for a case where two different objects are compared?

Reply:
oren says "We need to know the key to be able to check for and delete top level keys based on name", and this isn't too big of a deal

so i'm gonna resolve and merge, thank you!